### ISSUE TITLE
Rework the internal representation of Query from String to LinearSeqOptimized

### ISSUE THREAD

**Issue by bryce-anderson** [http4s/http4s#163]
Currently the ```Query``` is just a type alias for ```String```. It would be better as a real collection.

In order to remain true to the string form, it cannot be a simple map and must be a linear sequence.

**Comment by julien-truffaut** [comment]
what are the benefits to make `Query` part of the scala collection?

**Comment by bryce-anderson** [comment]
It allows use of simple operators such as ```++= mySeqOfKeys```, all the methods on the collections traits, etc. These may not be big wins. On the other hand, I see little downside.

**Comment by julien-truffaut** [comment]
Scala collections are often criticised because of all the inheritance and traits mixed in, it is difficult to get it right as methods are overriden in different places.

We already have an API for query in `Uri`, I think we could use the same in `Query`, e.g.

```
final class Query(...) {
  def +?[K: QueryParamKeyLike, T: QueryParamEncoder](name: K, value: T): Query = ...
}

case class Uri(...) {
  def +?[K: QueryParamKeyLike, T: QueryParamEncoder](name: K, value: T): Query =
    copy(query = query +?(name, value)) // you can even use lenses :p
}
```

**Comment by bryce-anderson** [comment]
+1 to using the api we have in ```Uri```. I'm in the process of making tests work right now, but once I'm done, I'll go about adding those methods.

I know some people don't like the scala collection library, but I don't subscribe to that concern. Our ```Headers``` collection also inherits from scala collections traits, and they also work nicely. My general stance is: if you don't want to use it as a scala collection, you don't have to. 

I'm also happy to include scalaz typeclass instances in the package object for those who would prefer to use those patterns.

**Comment by rossabaker** [comment]
This is now closed, no?

**Comment by bryce-anderson** [comment]
Yes, closed by #165.
